### DISCUSSION TITLE
Would it be helpful if there is an API to allow applications to export/show/output current running configuration?

### DISCUSSION THREAD

**Discussion by zbalkan** [dotnet/runtime#88017] [Ideas]
# Summary

An application can read configuration from registry, local files, environment variables, command line parameters, an external configuration source, or a database. However there are problems with the running configuration:

- System-wide configurations vs more specific configurations (per user, per app, per instance?)
- Multiple same level configuration resources overriding each other
- "first occurrence wins" vs "last occurrence wins"
- Multiple configuration resources for different use cases
- Environment variables passed through the parent process (Environment variables can be changed but the application needs a restart for the new values)

Would a common API like `show-params` of [click-extra](https://kdeldycke.github.io/click-extra/parameters.html#show-params-option) be helpful for devs who uses `Microsoft.Extensions.Configuration`?

It would be great if it would be possible to have a Cisco network devices `show running-config` command, which outputs the running config. Although one mostly types just `show run` and the device implies the rest of the command.

## Details

- It all started on the discussion here: https://news.ycombinator.com/item?id=36465886
- Multiple sources of configuration and multiple implementations depending on the software causes havoc
- With `Process Monitor/Explorer` or `strace` we can see which files and registry values are accessed. Also, it is possible to access environment variables but it is trickier on Linux.
- Without dumping the memory of a running application, can we not just print out the current configuration values in a structured way -preferable something machine-readably and hopefully human readable?

```json
{
    "configuration": [
        {
            "source": "registry",
            "path": "HKLM/...",
            "values": {
                "key": "value"
            }
        },
        {
            "source": "file",
            "path": "C:\\ProgramData\\...",
            "values": {
                "key": "value"
            }
        },
        {
            "source": "environment",
            "path": "...",
            "values": {
                "APP_SPECIFIC_ENV_VAR": "value"
            }
        }
    ]
}
```
Also, just providing this would provide an optional way. What does it take to standardize the usage is another issue.

**Comment by davidfowl** [comment]
Not sure it should be a command like that but there's an [API](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.configurationrootextensions.getdebugview?view=dotnet-plat-ext-7.0) for getting the debug view for a configuration.

@JamesNK also made [this change](https://github.com/dotnet/runtime/pull/86624) to make it easier to see where configuration comes from in the debugger.
